numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1120,6 +1120,30 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
         goto fail;
     }
 
+    if (!PyArray_CanCastArrayTo(indices, dtype, NPY_SAFE_CASTING)) {
+        if (PyArray_SIZE(indices) == 0) {
+            /*
+             * Unfortunately, if the input array is empty, then indices would
+             * likely have a default dtype of float64, which isn't compatibile.
+             * Force a recast of an empty array.
+             */
+            PyArrayObject *indices_tmp = NULL;
+            Py_INCREF(dtype);
+            indices_tmp = (PyArrayObject*)PyArray_FromArray(indices, dtype,
+                                                            NPY_ARRAY_FORCECAST);
```

Comment:
My argument about handling the second one is two-part. First, it is the path of least surprise:
```
np.unravel_index([], (2, 3))  # works
np.unravel_index(np.array([]), (2, 3))  # wtf? I hate numpy! rawr!
```
Second, one rarely actually uses unravel_index() and ravel_multi_index() in isolation. Usually, the inputs are coming from functions such as np.nonzero(). Now, while np.nonzero() does correctly set the dtype for empty arrays, are we sure other functions elsewhere in numpy does this? What about elsewhere in the scipy ecosystem?

Last night, I had a crazy idea for a wildcard dtype that would be used for the default dtype in empty arrays... haven't fleshed it out yet, though.